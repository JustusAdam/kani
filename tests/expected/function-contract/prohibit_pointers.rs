// Copyright Kani Contributors
// SPDX-License-Identifier: Apache-2.0 OR MIT
// kani-flags: -Zfunction-contracts

struct HidesAPointer(*mut u32);

#[kani::ensures(true)]
fn hidden_pointer(h: HidesAPointer) {}

#[kani::proof_for_contract(hidden_pointer)]
fn harness() {}

#[kani::ensures(true)]
fn plain_pointer(t: *mut i32) {}

#[kani::proof_for_contract(plain_pointer)]
fn plain_ptr_harness() {}

#[kani::ensures(true)]
fn return_pointer() -> *const usize {
    unreachable!()
}

#[kani::proof_for_contract(return_pointer)]
fn return_ptr_harness() {}

#[kani::ensures(true)]
fn allowed_pointer(t: *const bool) {}

#[kani::proof_for_contract(allowed_pointer)]
fn allowed_pointer_harness() {}

#[kani::ensures(true)]
fn allowed_mut_ref(t: &mut bool) {}

#[kani::proof_for_contract(allowed_mut_ref)]
fn allowed_mut_ref_harness() {}

#[kani::ensures(true)]
fn allowed_ref(t: &bool) {}

#[kani::proof_for_contract(allowed_ref)]
fn allowed_ref_harness() {}

#[kani::ensures(true)]
fn allowed_mut_return_ref<'a>() -> &'a mut bool {
    unreachable!()
}

#[kani::proof_for_contract(allowed_mut_return_ref)]
fn allowed_mut_return_ref_harness() {}
