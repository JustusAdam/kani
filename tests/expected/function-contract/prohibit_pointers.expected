error: Function argument contains a mutable pointer (*mut u32). This is prohibited for functions with contracts, as they cannot yet reason about the pointer behavior.

error: Function argument contains a mutable pointer (*mut i32). This is prohibited for functions with contracts, as they cannot yet reason about the pointer behavior.

error: Function argument contains a return pointer (*const usize). This is prohibited for functions with contracts, as they cannot yet reason about the pointer behavior.